### PR TITLE
Extend safeTemplateFilepath to cover more cases.

### PR DESCRIPTION
**Pull request by pstibrany**
#### What this PR does
This PR enhances `safeTemplateFilepath` check in alertmanager to cover more cases:

- template name `../tmpfile`, stored into `/tmp` dir
- empty template name
- template name being just `.`

All of these were previously accepted by `safeTemplateFilepath.`

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/alertmanager/multitenant_test.go [hunk 1]
```diff
@@ -1980,6 +2001,27 @@ func TestSafeTemplateFilepath(t *testing.T) {
 	}
 }
 
+func TestIsFilePathInsideDirectory(t *testing.T) {
+	assert.False(t, isFilePathInsideDirectory("/", "/"))
+	assert.False(t, isFilePathInsideDirectory("/", ""))
+	assert.True(t, isFilePathInsideDirectory("/", "/test"))
```
Comment (56quarters) [inline pkg/alertmanager/multitenant_test.go:2007]: nit: could the cases expected to be `false` and the cases expected to be `true` be grouped? Maybe sub-tests?

File: pkg/alertmanager/multitenant.go [hunk 2]
```diff
@@ -1262,16 +1262,42 @@ func safeTemplateFilepath(dir, templateName string) (string, error) {
...
+	if !isFilePathInsideDirectory(containerDir, actualPath) {
 		return "", fmt.Errorf("invalid template name %q: the template filepath is escaping the per-tenant local directory", templateName)
 	}
 
 	return actualPath, nil
 }
 
+// Returns true if file described by filePath is inside given directory. (Doesn't check for symbolic links)
+// Directory argument should not end with separator, ie. it must be filepath.Clean-ed.
+func isFilePathInsideDirectory(absDir string, absFilePath string) bool {
```
Comment (56quarters) [inline pkg/alertmanager/multitenant.go:1282]: nit: can you make the requirements of this function a little more clear? I.e. "(Doesn't check for symbolic links so the paths provided must already be absolute paths via filepath.Abs())" or similar?

**Review by 56quarters — APPROVED** [review]
LGTM, just a few (non-blocking) nits

File: pkg/alertmanager/multitenant.go [hunk 3]
```diff
@@ -1262,16 +1262,42 @@ func safeTemplateFilepath(dir, templateName string) (string, error) {
...
+	}
+
+	d := filepath.Dir(absFilePath)
+	if d == absDir {
+		return true
+	}
+
+	// If path was empty, its Dir is ".".
+	// If path was "/", its Dir is just "/".
+	if d == "." || d == "/" {
```
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1295]:
```suggestion
	if d == "." || d == os.PathSeparator {
```

File: pkg/alertmanager/multitenant.go [hunk 4]
```diff
@@ -1262,16 +1262,42 @@ func safeTemplateFilepath(dir, templateName string) (string, error) {
...
 	// escape its own directory on disk.
-	if !strings.HasPrefix(actualPath, containerDir) {
+	if !isFilePathInsideDirectory(containerDir, actualPath) {
 		return "", fmt.Errorf("invalid template name %q: the template filepath is escaping the per-tenant local directory", templateName)
 	}
 
 	return actualPath, nil
 }
 
+// Returns true if file described by filePath is inside given directory. (Doesn't check for symbolic links)
```
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1280]:
```suggestion
// Returns true if file described by absFilePath is inside given directory. (Doesn't check for symbolic links)
```

File: pkg/alertmanager/multitenant.go [hunk 5]
```diff
@@ -1262,16 +1262,42 @@ func safeTemplateFilepath(dir, templateName string) (string, error) {
...
+// Directory argument should not end with separator, ie. it must be filepath.Clean-ed.
```
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1281]:
```suggestion
// absDir argument should not end with separator, ie. it must be filepath.Clean-ed.
```

File: pkg/alertmanager/multitenant.go [hunk 6]
```diff
@@ -1262,16 +1262,42 @@ func safeTemplateFilepath(dir, templateName string) (string, error) {
...
 
+	// If actualPath is same as containerDir, it's likely that actualPath was empty, or just ".".
+	if containerDir == actualPath {
+		return "", fmt.Errorf("invalid template name %q", templateName)
+	}
+
 	// Ensure the actual path of the template is within the expected directory.
 	// This check is a counter-measure to make sure the tenant is not trying to
 	// escape its own directory on disk.
-	if !strings.HasPrefix(actualPath, containerDir) {
```
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:1268]: Wouldn't the fix as easy as checking for:

```
if !strings.HasPrefix(actualPath, fmt.Sprintf("%s%c", containerDir, os.PathSeparator)) {
```

**Review by pracucci — APPROVED** [review]
Good job spotting it! I don't see any issue with your implemention (just left few nits) but I'm wondering if `isFilePathInsideDirectory` is too much engineered: I left a comment with a possible 1-line change alternative (would that cover all cases?)

**Comment by pstibrany** [comment]
> Good job spotting it! I don't see any issue with your implemention (just left few nits) but I'm wondering if `isFilePathInsideDirectory` is too much engineered: I left a comment with a possible 1-line change alternative (would that cover all cases?)

You are right, let's use simplified version.

(Nit: if we ever support Mimir on Windows, our unit tests will fail)

**Review by pracucci — APPROVED** [review]

**Pull request merged**